arch/risc-v/src/common/espressif/esp_qencoder.c

```diff
@@ -0,0 +1,593 @@
+                     priv->config->ch0_pulse_sig, 0);
+  esp_gpio_matrix_in(priv->config->ch1_gpio,
+                     priv->config->ch0_ctrl_sig, 0);
+
+  esp_gpio_matrix_in(priv->config->ch1_gpio,
+                     priv->config->ch1_pulse_sig, 0);
+  esp_gpio_matrix_in(priv->config->ch0_gpio,
+                     priv->config->ch1_ctrl_sig, 0);
+
+  /* Clear the Reset bit to enable the Pulse Counter */
+
+  pcnt_ll_clear_count(priv->dev, priv->config->pcntid);
+  pcnt_ll_start_count(priv->dev, priv->config->pcntid);
+
+  spin_unlock_irqrestore(&priv->lock, flags);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp_shutdown
+ *
+ * Description:
+ *   This method is called when the driver is closed.  The lower half driver
+ *   should stop data collection, free any resources, disable timer hardware,
+ *   and put the system into the lowest possible power usage state
+ *
+ ****************************************************************************/
+
+static int esp_shutdown(struct qe_lowerhalf_s *lower)
+{
+  struct esp_lowerhalf_s *priv = (struct esp_lowerhalf_s *)lower;
+
+  /* Disable PCNT clock */
+
+  periph_module_disable(PERIPH_PCNT_MODULE);
+
+  /* Make sure initial position is 0 */
+
+  priv->position = 0;
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp_position
+ *
+ * Description:
+ *   Return the current position measurement.
+ *
+ ****************************************************************************/
+
+static int esp_position(struct qe_lowerhalf_s *lower, int32_t *pos)
+{
+  struct esp_lowerhalf_s *priv = (struct esp_lowerhalf_s *)lower;
+  irqstate_t flags;
+  int32_t position;
+  int16_t count;
+
+  DEBUGASSERT(lower && priv->inuse);
+
+  flags = spin_lock_irqsave(&priv->lock);
+
+  position = priv->position;
+  count = (int16_t)(pcnt_ll_get_count(priv->dev, priv->config->pcntid) & 0xffff);
+
+  /* Return the position measurement */
+
+  *pos = position + count;
+
+  spin_unlock_irqrestore(&priv->lock, flags);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp_setposmax
+ *
+ * Description:
+ *   Set the maximum encoder position.
+ *
+ ****************************************************************************/
+
+static int esp_setposmax(struct qe_lowerhalf_s *lower, uint32_t pos)
+{
+  return -ENOTTY;
+}
+
+/****************************************************************************
+ * Name: esp_reset
+ *
+ * Description:
+ *   Reset the position measurement to zero.
+ *
+ ****************************************************************************/
+
+static int esp_reset(struct qe_lowerhalf_s *lower)
+{
+  struct esp_lowerhalf_s *priv = (struct esp_lowerhalf_s *)lower;
+  irqstate_t flags;
+  uint32_t regval;
+
+  sninfo("Resetting position to zero\n");
+  DEBUGASSERT(lower && priv->inuse);
```

Comment:
You can move DEBUGASSERT before setting *priv

Reply:
done